From dc60b9230fe32cf73598d7a0f83c501ed0d86e74 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 11 Feb 2025 11:15:51 +0800
Subject: [PATCH 156/206] mtk: hostapd: allow only new association to enter WPA
 state machine

If multiple association requests are received and thus multiple
association response are sent, the WPA state machine will be triggered m
ultiple times, which fails the 4-way handshake.

This commit allows only entering the WPA state machine only for STA's
first ascoiation.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 54c3cddaf..f031f8ad2 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7212,7 +7212,7 @@ static void handle_assoc_cb(struct hostapd_data *hapd,
 
 	if (sta->auth_alg == WLAN_AUTH_FT)
 		wpa_auth_sm_event(sta->wpa_sm, WPA_ASSOC_FT);
-	else
+	else if (new_assoc)
 		wpa_auth_sm_event(sta->wpa_sm, WPA_ASSOC);
 	hapd->new_assoc_sta_cb(hapd, sta, !new_assoc);
 	ieee802_1x_notify_port_enabled(sta->eapol_sm, 1);
-- 
2.45.2

